cpp/src/arrow/compute/memory_resources.cc

```diff
@@ -0,0 +1,155 @@
+
+MemoryResources::~MemoryResources() {}
+
+std::unique_ptr<MemoryResources> MemoryResources::Make() {
+  return std::unique_ptr<MemoryResources>(new MemoryResources());
+}
+
+Status MemoryResources::AddMemoryResource(std::shared_ptr<MemoryResource> resource) {
+  auto level = static_cast<size_t>(resource->memory_level());
+  if (stats_[level] != nullptr) {
+    return Status::KeyError("Already have a resource type registered with name: ",
+                            resource->ToString());
+  }
+  stats_[level] = std::move(resource);
+  return Status::OK();
+}
+
+size_t MemoryResources::size() const { return stats_.size(); }
+
+Result<MemoryResource*> MemoryResources::memory_resource(MemoryLevel memory_level) const {
+  auto level = static_cast<size_t>(memory_level);
+  if (stats_[level] == nullptr) {
+    return Status::KeyError("No memory resource registered with level: ",
+                            MemoryLevelName(memory_level));
+  }
+  return stats_[level].get();
+}
+
+std::vector<MemoryResource*> MemoryResources::memory_resources() const {
+  std::vector<MemoryResource*> arr;
+  for (auto&& resource : stats_) {
+    if (resource != nullptr) {
+      arr.push_back(resource.get());
+    }
+  }
+  return arr;
+}
+
+namespace {
+
+size_t GetTotalMemorySize() {
+#ifdef __APPLE__
+  int mib[2];
+  size_t physical_memory;
+  size_t length;
+  // Get the Physical memory size
+  mib[0] = CTL_HW;
+  mib[1] = HW_MEMSIZE;
+  length = sizeof(size_t);
+  sysctl(mib, 2, &physical_memory, &length, NULL, 0);
+  return physical_memory;
+#elif defined(_WIN32)
+  MEMORYSTATUSEX status;
+  status.dwLength = sizeof(status);
+  GlobalMemoryStatusEx(&status);
+  return status.ullTotalPhys;
+#else  // Linux
+  struct sysinfo si;
+  sysinfo(&si);
+  return (size_t)si.freeram;
+#endif
+}
+
+struct CPUMemoryResource : public MemoryResource {
+  explicit CPUMemoryResource(arrow::MemoryPool* pool, float memory_limit_threshold = 0.75)
+      : MemoryResource(MemoryLevel::kCpuLevel), pool_(pool) {
+    total_memory_size_ = GetTotalMemorySize();
+    memory_limit_ =
+        static_cast<int64_t>(std::round(memory_limit_threshold * total_memory_size_));
+  }
+
+  int64_t memory_used() override { return pool_->bytes_allocated(); }
+
+  int64_t memory_limit() override { return memory_limit_; }
+
+ private:
+  arrow::MemoryPool* pool_;
+  int64_t memory_limit_;
+  int64_t total_memory_size_;
+};
+
+static std::unique_ptr<MemoryResources> CreateBuiltInMemoryResources(MemoryPool* pool) {
+  auto resources = MemoryResources::Make();
+
+  // CPU MemoryLevel
+  auto cpu_level = std::make_shared<CPUMemoryResource>(pool);
+  DCHECK_OK(resources->AddMemoryResource(std::move(cpu_level)));
+
+  // Disk MemoryLevel ...
+
+  return resources;
+}
+
+}  // namespace
+
+MemoryResources* GetMemoryResources(MemoryPool* pool) {
+  static auto resources = CreateBuiltInMemoryResources(pool);
+  return resources.get();
+}
```

Comment:
In the other places where we have this "process-wide default" we use a singleton pattern.  Do we need a process-wide default here?  If so, should we use the singleton pattern?

If we don't want a process-wide default then what is the difference between this method and `MemoryResources::Make`?